padertorch/modules/normalization.py

```diff
@@ -1,104 +1,337 @@
+                self.beta = nn.Parameter(
+                    torch.zeros(reduced_shape), requires_grad=True
+                )
+            else:
+                self.beta = None
         else:
-            self.register_parameter('weight', None)
-            self.register_parameter('bias', None)
-        self.reset_parameters()
+            self.gamma = None
+            self.beta = None
 
-    def _check_input_dim(self, input):
-        assert input.ndim >= self.statistics_axis
+    @property
+    def runnning_var(self):
+        n = torch.max(self.num_tracked_values, 2. * torch.ones_like(self.num_tracked_values))
+        running_var = self.running_power
+        if self.shift:
+            running_var = n / (n-1) * (running_var - self.running_mean ** 2)
+        running_var = running_var + self.eps
+        assert (running_var >= 0).all(), running_var.min()
+        return running_var
+
+    def reset_running_stats(self):
+        if self.track_running_stats:
+            self.num_tracked_values.zero_()
+            if self.shift:
+                self.running_mean.zero_()
+            if self.scale:
+                self.running_power.fill_(1)
 
     def reset_parameters(self):
-        if self.affine:
-            torch.nn.init.uniform_(self.weight)
-            torch.nn.init.zeros_(self.bias)
-
-    def get_statistics(self, tensor):
-        if self.order == 'l1':
-            mean = sequence_reduction(
-                torch.mean,
-                tensor,
-                axis=self.statistics_axis,
-                keepdims=True
-            )
-            norm = sequence_reduction(
-                torch.mean,
-                torch.abs(tensor),
-                axis=self.statistics_axis,
-                keepdims=True
-            )
-        elif self.order in ['l2', 'mean']:
-            mean = sequence_reduction(
-                torch.mean,
-                tensor,
-                axis=self.statistics_axis,
-                keepdims=True
-            )
-            norm = sequence_reduction(
-                torch.mean,
-                torch.abs(tensor - mean)**2,
-                axis=self.statistics_axis,
-                keepdims=True
+        self.reset_running_stats()
+        if self.gamma is not None:
+            nn.init.ones_(self.gamma.scale)
+        if self.beta is not None:
+            nn.init.zeros_(self.beta.shift)
+
+    def forward(self, x, sequence_lengths=None):
+        if self.training or not self.track_running_stats:
+            y, mean, power, n_values = normalize(
+                x, gamma=self.gamma, beta=self.beta,
+                statistics_axis=self.statistics_axis,
+                batch_axis=self.batch_axis, sequence_axis=self.sequence_axis,
+                sequence_lengths=sequence_lengths, shift=self.shift, scale=self.scale,
+                eps=self.eps
             )
+            if self.track_running_stats:
+                self.num_tracked_values += n_values.data
```

Comment:
Are these `.data`s here (and in the following code) remainders of pre-PyTorch 0.4 code? If so, it is safer to replace it with `.detach()` (see https://pytorch.org/blog/pytorch-0_4_0-migration-guide/#what-about-data). If not, I would be interested in why you use `.data` here.